File: src/0081.Search_in_Rotated_Sorted_Array2/search_in_rotated_sorted_array2.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int ll = 0;
        int rr = nums.size() - 1;
        int mid;
        while(ll <= rr){
            mid = ll + (rr - ll) / 2;
            if(nums[mid] == target){
                return true;
            }
            if(nums[ll] == nums[rr] && nums[ll] == nums[mid]){
                ll++;
                rr--;
            }
            else if(nums[ll] <= nums[mid]){
                if(target >= nums[ll] && target <= nums[mid]){
                    rr = mid - 1;
                }
                else{
                    ll = mid + 1;
                }
            }
            else{
                if(target >= nums[mid] && target <= nums[rr]){
                    ll = mid + 1;
                }
                else{
                    rr = mid - 1;
                }
            }
        }
        return false;
    }
};

File: src/0081.Search_in_Rotated_Sorted_Array2/search_in_rotated_sorted_array2.java
class Solution {
    public boolean search(int[] nums, int target) {
        if(nums == null || nums.length == 0){
            return false;
        }
        int left = 0;
        int right = nums.length - 1;
        while(left < right){
            int mid = left + (right - left + 1) / 2;
            if(nums[mid] < nums[right]){
                if(nums[mid] <= target && target <= nums[right]){
                    left = mid;
                } else{
                    right = mid - 1;
                }
            } else if(nums[mid] > nums[right]){
                if(nums[left] <= target && target <= nums[mid - 1]){
                    right = mid - 1;
                } else{
                    left = mid;
                }
            } else{
                if(nums[right] == target){        //!!!!!!!!!!!!!!!!!!!
                    return true;
                }
                right--;
            }
        }
        if(nums[left] == target){
            return true;
        }
        return false;
    }
}